### DISCUSSION TITLE
Python script for bike trip

### DISCUSSION THREAD

**Discussion by Shohreh** [organicmaps/organicmaps#1315] [Ideas]
Hello,

I'm no developer, and wrote this little Python script to edit KML files before leaving for a bike trip.

The script 1) adds/edits the top-most "name" using the filename, 2) colorizes tracks in fat red and/or colorizes waypoints using the fifteen other colors availables as an easy way to show multiple layers and still know what kind of location it is (housing, restaurants, bikeshops, etc.)

```
import lxml.etree as et
from lxml import html
import sys
import os
from glob import glob

#Colorize tracks in red
LS = "<Style><LineStyle><color>FF0000FF</color><width>6</width></LineStyle></Style>"
#Red not available, because already used as default for waypoints in track file
colors = ["#placemark-yellow", "#placemark-blue", "#placemark-green", "#placemark-purple", "#placemark-brown", "#placemark-pink", "#placemark-deeppurple","#placemark-lightblue","#placemark-cyan","#placemark-orange", "#placemark-teal","#placemark-lime","#placemark-deeporange","#placemark-gray","#placemark-bluegray"]

#============= Read file(s) from command line; Windows and *nix shells don't behave the same way
def argGlob(args=None):
	if args is None: args = sys.argv
	
	return [subglob for arg in args for subglob in glob(arg)]

#Exit if no input file on command line
arguments = len(sys.argv) - 1
if arguments < 1:
	print ("Must pass at least one KML file, or *.KML")
	sys.exit()

#=============
counter=0
for item in argGlob(sys.argv[1:]):
	print(f"**************** Handling {item}")

	BASENAME, EXTENSION = os.path.splitext(os.path.basename(item))
	OUTPUTFILE = f"{BASENAME}.EDITED{EXTENSION}"
	parser = et.XMLParser(remove_blank_text=True,strip_cdata=False)
	tree = et.parse(item,parser)
	root = tree.getroot()
	
	#=========== Remove namespace prefixes
	#https://stackoverflow.com/questions/60486563/remove-namespace-from-xml-with-comment-python
	for elem in root.getiterator():
		# For elements, replace qualified name with localname
		if not(type(elem) == et._Comment):
			elem.tag = et.QName(elem).localname

		# Remove attributes that are in a namespace
		for attr in elem.attrib:
			if "{" in attr:
				elem.attrib.pop(attr)	
	# Remove unused namespace declarations
	et.cleanup_namespaces(root)

	#remove comments
	et.strip_tags(tree,et.Comment)

	#=========== Remove blocs/elements
	for el in root.iter('snippet', 'LookAt', 'Style', 'StyleMap', 'tessellate', 'ExtendedData'):
		parent = el.getparent()
		parent.remove(el)
	
	#=========== Edit/insert top-most name
	r = root.find('./Document/name')
	if r is not None:
		print("Name exists: ",r.text)
		r.text = BASENAME
	else:
		print("Name does not exist")
		name = et.Element("name")
		name.text = BASENAME
		r= root.find("./Document")
		r.insert(0, name)

	#=========== If file contains one or more tracks, just redify tracks and ignore waypoints
	#if tree contains at least one track
	if root.find('.//LineString') is not None:
		print("One or more track")
		for el in root.iter('LineString'):
			el.addprevious(et.XML(LS))
	#if file only contains waypoints, colorize them using current color
	else:
		print("Only waypoints")
		print("Using ",colors[counter])

		#just to know
		#placemark-yellow
		currcol = colors[counter].split('-')[1]
		OUTPUTFILE = f"{BASENAME}.{currcol}.EDITED{EXTENSION}"

		if root.find('.//styleUrl') is not None:
			for styleUrl in root.iter('styleUrl'):
				print("Found styleUrl: ",styleUrl.text)
				styleUrl.text = colors[counter]
		else:
			print("No styleUrl found")
			for pm in root.iter('Placemark'):
				et.SubElement(pm, "styleUrl").text = colors[counter]
				pass
				
		counter += 1
		#Only 15 colors available: If more than 15 files, start again
		if counter == len(colors):
			counter=0

	#Write modified tree to new file
	tree.write(OUTPUTFILE, pretty_print=True)

```
Cheers,

**Comment by rtsisyk** [comment]
Thanks! ---- On Sat, 18 Sep 2021 15:32:00 +0300  ***@***.***  wrote ----
Hello,
I'm no developer, and wrote this little Python script to edit KML files before leaving for a bike trip.
The script 1) adds/edits the top-most "name" using the filename, 2) colorizes tracks in fat red and/or colorizes waypoints using the fifteen other colors availables as an easy way to show multiple layers and still know what kind of location it is (housing, restaurants, bikeshops, etc.)
import lxml.etree as et
from lxml import html
import sys
import os
from glob import glob

#Colorize tracks in red
LS = "<Style><LineStyle><color>FF0000FF</color><width>6</width></LineStyle></Style>"
#Red not available, because already used as default for waypoints in track file
colors = ["#placemark-yellow", "#placemark-blue", "#placemark-green", "#placemark-purple", "#placemark-brown", "#placemark-pink", "#placemark-deeppurple","#placemark-lightblue","#placemark-cyan","#placemark-orange", "#placemark-teal","#placemark-lime","#placemark-deeporange","#placemark-gray","#placemark-bluegray"]

#============= Read file(s) from command line; Windows and *nix shells don't behave the same way
def argGlob(args=None):
	if args is None: args = sys.argv
	
	return [subglob for arg in args for subglob in glob(arg)]

#Exit if no input file on command line
arguments = len(sys.argv) - 1
if arguments < 1:
	print ("Must pass at least one KML file, or *.KML")
	sys.exit()

#=============
counter=0
for item in argGlob(sys.argv[1:]):
	print(f"**************** Handling {item}")

	BASENAME, EXTENSION = os.path.splitext(os.path.basename(item))
	OUTPUTFILE = f"{BASENAME}.EDITED{EXTENSION}"
	parser = et.XMLParser(remove_blank_text=True,strip_cdata=False)
	tree = et.parse(item,parser)
	root = tree.getroot()
	
	#=========== Remove namespace prefixes
	#https://stackoverflow.com/questions/60486563/remove-namespace-from-xml-with-comment-python
	for elem in root.getiterator():
		# For elements, replace qualified name with localname
		if not(type(elem) == et._Comment):
			elem.tag = et.QName(elem).localname

		# Remove attributes that are in a namespace
		for attr in elem.attrib:
			if "{" in attr:
				elem.attrib.pop(attr)	
	# Remove unused namespace declarations
	et.cleanup_namespaces(root)

	#remove comments
	et.strip_tags(tree,et.Comment)

	#=========== Remove blocs/elements
	for el in root.iter('snippet', 'LookAt', 'Style', 'StyleMap', 'tessellate', 'ExtendedData'):
		parent = el.getparent()
		parent.remove(el)
	
	#=========== Edit/insert top-most name
	r = root.find('./Document/name')
	if r is not None:
		print("Name exists: ",r.text)
		r.text = BASENAME
	else:
		print("Name does not exist")
		name = et.Element("name")
		name.text = BASENAME
		r= root.find("./Document")
		r.insert(0, name)

	#=========== If file contains one or more tracks, just redify tracks and ignore waypoints
	#if tree contains at least one track
	if root.find('.//LineString') is not None:
		print("One or more track")
		for el in root.iter('LineString'):
			el.addprevious(et.XML(LS))
	#if file only contains waypoints, colorize them using current color
	else:
		print("Only waypoints")
		print("Using ",colors[counter])

		#just to know
		#placemark-yellow
		currcol = colors[counter].split('-')[1]
		OUTPUTFILE = f"{BASENAME}.{currcol}.EDITED{EXTENSION}"

		if root.find('.//styleUrl') is not None:
			for styleUrl in root.iter('styleUrl'):
				print("Found styleUrl: ",styleUrl.text)
				styleUrl.text = colors[counter]
		else:
			print("No styleUrl found")
			for pm in root.iter('Placemark'):
				et.SubElement(pm, "styleUrl").text = colors[counter]
				pass
				
		counter += 1
		#Only 15 colors available: If more than 15 files, start again
		if counter == len(colors):
			counter=0

	#Write modified tree to new file
	tree.write(OUTPUTFILE, pretty_print=True)


Cheers,
